### PR TITLE
Dedicated image for the Lenovo Thinkpad X13s

### PR DESCRIPTION
**Pull request by jglathe**
This is sort of experimental, me trying to get a handle on how this works. The changes are v1.0, still with local dependencies, I will put up all of it into github repos, after creating some order. There are a few things noteworthy, though.

1. Lenovo Thinkpad X13s is an aarch64 target, you need a device tree to successfully boot it. There seems to be no real mechanism to detect / select / provide this from the ISO yet.  There is a (beta) mechanism in the UEFI BIOS to load the dtb from the ESP and provide it to the kernel, and this works (with Ubuntu at least). But this is outside of the ISO, and these boxes aren't shipped with the dtb. Future Snapdragon X Elite based boxes may be, but there is still the issue that there is no standard way on how this may work (yet).
2. You also need boot-critical drivers to load first from the initramfs, and certain firmwares (for the coprocessors, adsp and cdsp, gpu) need to be in the initramfs, too. 
3. I (hopefully) understood the boot part so far that it is using parts of grub-efi-aarch64 to bring it up from the iso. My experience with trying to use systemd-boot with these uefi BIOSes (even though we have efivars now on x13s) is rather poor. The installed system on an external SSD, when connected to the laptop, will cause it not to boot. I can boot when I remove the ESP, and boot via an already existing grub. I do so on a Windows Dev Kit 2023, same SoC. Maybe going grub for aarch64 until this works is worth a consideration.
4. You need specific kernel options to make it boot due to the incompleteness of some hardware definitions. AFAIU this is not really an exception.
5. I used my own deb packages for the kernel, but you can do this with the pop-os_linux tree too, but you need different kernel options for it to work.

In conclusion this means that you maybe can build an ISO to boot for these targets, but its not generic at all. At least yet. So branches / configurations for specific targets may be the way to get something for the time being.

### CODE REVIEW

File: mk/iso.mk [hunk 1]
```diff
@@ -14,18 +14,23 @@ $(BUILD)/iso_casper.tag: $(BUILD)/live $(BUILD)/chroot.tag $(BUILD)/live.tag $(B
 	# Create new casper directory
 	mkdir -p "$(BUILD)/iso/$(CASPER_PATH)"
 
+ifeq ($(DISTRO_MACHINE),x13s)
+	# copy over dtb to casper
+	cp -v "$(BUILD)/live/usr/lib/linux-image-6.9.1+/qcom/sc8280xp-lenovo-thinkpad-x13s.dtb" "$(BUILD)/iso/$(CASPER_PATH)/$(DISTRO_MACHINE_DTB)"
```
Comment (thefossguy) [inline mk/iso.mk:19]: Why not copy _all_ DTBs? That way, more ARM platforms that boot using DT over ACPI. I know that NixOS and Rocky Linux currently boot like this: U-Boot -> GRUB -> Linux. While you really only need the FDT in U-Boot and can boot without the in-kernel DTB, it's not recommended.

Jeremy and Aron: Sorry for hijacking this :P

Comment (jglathe) [reply mk/iso.mk:19]: Hmm how do you select the right one when its available, you need to do so before loading linux.

Comment (jglathe) [reply mk/iso.mk:19]: Oh look @travmurav seems to be on it https://github.com/TravMurav/dtbloader

Comment (thefossguy) [reply mk/iso.mk:19]: You just place all DTBs in `/boot/dtb` and the bootloader picks it up.

Comment (jglathe) [reply mk/iso.mk:19]: > You just place all DTBs in `/boot/dtb` and the bootloader picks it up.

... which one? OTOH, as described, initramfs is a bit special (won't boot, or only from nvme, if the modules and firmware hooks aren't there). For me this sounds a lot like you still need a device specific layer of code, even if we manage to get something like dtbloader figures out the right dtb. But this would be doable for one image, at least.

Comment (thefossguy) [reply mk/iso.mk:19]: Putting the DTBs in the necessary paths is _already_ handled by the .deb package of the kernel. This shouldn't be done manually.

Comment (jglathe) [reply mk/iso.mk:19]: ... if it works. It depends on flash-kernel, and my impression was that it doesn't, haven't found the cause yet. Therefore the direct command as a workaround / hack, as the whole thing currently is.

**Comment by jackpot51** [comment]
I am re-opening this because I think there are some things we can do to improve the situation for ARM devices, while still hoping for complete upstream kernel support.

In particular, our ISOs are missing the devicetree files from the included kernel. We should not just include the files from one device (the X13s), but instead should include all of the files, and ensure that the bootloader correctly loads them.

**Comment by jglathe** [comment]
Which opens the topic discussed [above](https://github.com/pop-os/iso/pull/330#discussion_r1620809964). I guess you need a few things:

- detection of machine
- specific initramfs support (for qcom especially as it seems)
- boot parameters for the grub entry 
- kernel config that has all required features / modules on

It will probably be an incremental work going from device(class) to device. I have two pretty similar boxes, the X13s, and the Windows Dev Kit 2023. They boot with the same kernel, modules list and config, have separate dtbs. And a part of the firmware is signed device-specific.

**Comment by jglathe** [comment]
flash-kernel works with dtb installation on the 24.04 install, btw. Haven't tried to boot anew with systemd-boot yet.

**Comment by jackpot51** [comment]
Now that the noble branch is merged, this will have to be reopened on the master branch.

**Comment by jglathe** [comment]
With an updated version. For boot-only (and install) purposes you can forego all firmwares except for the gpu firmware, and one needs a modules list. Then it boots on sc8280xp and x1e80100. With [dtbloader](https://github.com/TravMurav/dtbloader) it might be feasible to actually boot by arm64 machine.
